gix-worktree/src/index/status/worktree.rs

```diff
@@ -0,0 +1,272 @@
+    ) -> Result<Status<T>, Error> {
+        // TODO fs caache
+        let worktree_path = path::try_from_bstr(git_path).map_err(|_| Error::IllformedUtf8)?;
+        let worktree_path = self.worktree.join(worktree_path);
+        let metadata = match worktree_path.symlink_metadata() {
+            // TODO: check if any parent directory is a symlink
+            //       we need to use fs::Cache for that
+            Ok(metadata) if metadata.is_dir() => {
+                // index entries are normally only for files/symlinks
+                // if a file turned into a directory it was removed
+                // the only exception here are submodules which are
+                // part of the index despite being directories
+                //
+                // TODO: submodules:
+                //   if entry.mode.contains(Mode::COMMIT) &&
+                //     resolve_gitlink_ref(ce->name, "HEAD", &sub))
+                return Ok(Status::Removed);
+            }
+            Ok(metadata) => metadata,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(Status::Removed),
+            Err(err) => {
+                return Err(err.into());
+            }
+        };
+        if self.options.check_added && entry.flags.contains(index::entry::Flags::INTENT_TO_ADD) {
+            return Ok(Status::Added);
+        }
+        let new_stat = index::entry::Stat::from_fs(&metadata)?;
+        let executable_bit_changed =
+            match entry
+                .mode
+                .change_to_match_fs(&metadata, self.options.fs.symlink, self.options.fs.executable_bit)
+            {
+                Some(index::entry::mode::Change::Type { .. }) => return Ok(Status::TypeChange),
+                Some(index::entry::mode::Change::ExecutableBit) => true,
+                None => false,
+            };
+
+        // Here we implement racy-git. See racy-git.txt in the git documentation for a detailed documentation
+        // A file is racy if:
+        // * it's mtime is at/after the last index timestamp, it's entry stat information
+        //   matches the on-disk file but the file contents are actually modified
+        // * it's size is 0 (set after detecting a file was racy previously)
+        //
+        // The first case is detected below (by checking the timestamp if the file is marekd umodified).
+        // The second case is usually detected either because the on-disk file
+        // is not empty (hence the basic stat match fails) or by checking
+        // whether the size doesn't fit the oid (the oid of an empty blob is a constant)
+        let mut racy_clean = false;
+        if !executable_bit_changed
+            && new_stat.matches(&entry.stat, self.options.stat)
+            && (!entry.id.is_empty_blob() || entry.stat.size == 0)
+        {
+            racy_clean = new_stat.is_racy(self.timestamp, self.options.stat);
+            if !racy_clean {
+                return Ok(Status::Unchanged);
```

Comment:
Can this get some explanation in code? My intuition here was the inverse, but even then I don't get why this has to be like that. It's probably all explained above, yet I beg for more 😅.
*(I messed with this part to see if a test breaks, but since this didn't happen I think it's even more important to explain it, or find a way to add a test)*

Reply:
I hoped that the comment above would explain this but it is admittedly a bit tricky to understand and I always struggle with prose so this is probably not a great explanation.

Basically the racy detection is a safety mechanism that ensures we can always just compare the stat information between index and worktree and if they match we don't need to look at the content. This usually just works but if a file updates quickly we could run into the following situation:

* save file version A from disk into worktree (git add)
* file is changed so fast that the mtime doesn't change (we only looks at seconds by default)
* file contents change but filesize stays the same (so "foo" -> "bar" has the same size but different content)

Now both mtime and size (and all other stat information) is the same but the file has actually changed. This case is called racily clean. The file should show up as changed but due to a data race it doesn't. This is the racy git problem.

To solve this we do the following trick: Whenever we modify the index (which includes git status) we save the current timestamp before the modification starts. This timestamp fundamentally represents a checkpoint of sorts. We "promise" ourselves that after the modification finishes all entries modified before this timestamp have the racy git problem resolved.

So now when we modify the index we must resolve the racy git problem somehow. To do only need to look at unchanged entries. Changed entries are not interesting since they are already showing up as changed anyway so there isn't really a racecondition to worry about. This also explains why removing the return here doesn't have an apparent effect. This entire branch here is just the optimization of "don't even look at index entries where the stat hasn't changed". If we don't have this optimization the result shouldn't change, our status implementation will just be super slow :D


We calculate whether this change is `racy_clean` (so if the last timestamp is before or the same as the mtime of the entry which is what `new_stat.is_racy(..)` does) in the branch and only if we are sure that there is no race condition do we return early.

I hope that's clearer now the race condition stuff is quite subtle :D

Reply:
Thank you so much! The additional explanation I formatted and placed as function documentation, just to salvage the time that went into the long form :).

Something I *still* struggle with is the logic around empty_blob hashes and zero size, but now that there is the new test it's easy to change things (maybe the way I think they should be) and find tests turn red. That's exactly how it should be.

I did manage to keep tests passing with this…
```
        // && (!entry.id.is_empty_blob() || entry.stat.size == 0)
```
…but I really wasn't able to find a test that would fail :/.

I also just saw that this 'trick' with marking entries `size = 0 but non-empty hash` helps deal with filtering, too.

That said, once I looked at the tests I am sure it will finally click.

Lastly, something I also *didn't* understand is how the opposite case is handled. A file seems to have changed, but it actually didn't due to raciness. And the answer is that if it seems to have changed, we always validate by checking the content.

I also added the implementation for allowing ODB access, and in the course of it learned why `LazyBlob` was made to  be usable only once :). It's smart, I never thought of doing that to workaround the lifetime issues you get when trying to return data from an FnMut closure.

Reply:
> I did manage to keep tests passing with this…

  >      // && (!entry.id.is_empty_blob() || entry.stat.size == 0)

> …but I really wasn't able to find a test that would fail :/.

Ah sorry about that I don't have a special test case for empty files yet. That will be a bit harder to test I need to think about that. I don't want to interrupt your process so I could that as a followup.


You are exactly right the racy git problem is only about racily dirty files. Files where the stat changed in any way always have their content checked there is no way around that. But that is also not really a race condition at all since it's perfectly valid to overwrite a file with content `foo` with `bar` and the overwrite it with `foo` again. The mtime/stat has definitely changed but the file content hasn't. 

In other words: If the stat is the same that means the file is unchanged, if the stat is not the same we don't know anything and just need to check the file.  The racy git problem is that the first assumption is also not true if files change too quickly which is why you also need to check the file at least once (if the mtime is newer than the index timestamp).

This leads perfectly to the size == 0 case. Now after we have checked once and determined that the file actually did change despite the stat information saying it didn't we need to purposefully messup our stat information somehow to ensure that next time the stat comparison failed (as the timestamp will be updated and the `is_racy` function would then return false). To do that we set the filesize to 0.

You are actually right that this check is needed right here. I was a bit confused because git has a special check but for filesize 0 but now I think this just comes from the fact that their code is structured a bit different. This check doesn't hurt in any way it just doesn't have an effect so please remove it.  The key realization here (which ( missed previously) is that empty blobs can not be racily clean. All empty files are identical. For an empty file to become dirty its filesiez would need to change (which again would make the stat comparison fail).

However, there is a similar check is needed in the content comparison that is actually needed:

``` rust
        if entry.stat.size as usize != blob_size && (entry.id.is_empty_blob() || entry.stat.size != 0) {}
```)

This is a fast path so we don't actually load the file from disk when we already know the file size is different. However for racily smudged entries (so entries where we set the filesize to 0 as explained above) this check always yields that the file has changed even when it really did not change, because the saved filesize is wrong. So we disable the fast path if the file size is equal to 0 :
 
``` rust
        if entry.stat.size as usize != blob_size &&  entry.stat.size != 0 {}
```

This usually works however there is one edge-case: Empty blobs would always fail the second check but for them `entry.stat.size` being 0 is correct. So we need an extra condition to enable the fastpath for them which leads to the condition as implemented.

Reply:
>  That will be a bit harder to test I need to think about that. I don't want to interrupt your process so I could that as a followup.
> This check doesn't hurt in any way it just doesn't have an effect so please remove it.

Yes, that would be nice as a follow-up, whether explicit or as part of the next round. Maybe then once there is a test you can find a reason to re-add it, maybe it can safe some time in case of a race. Thinking about it, when filters are involved I think the race-conditions are met much more often and I have a feeling that this test is important. So, I think I will go with a TODO then, but you can remove it if it really turns out to be useless (again, I fear filters a bit)

And thanks again for the additional explanation, maybe I can adapt some of that as in-code comment, too :).